hypothesis-python/src/hypothesis/searchstrategy/collections.py

```diff
@@ -137,9 +139,26 @@ def do_draw(self, data):
         )
         seen = set()
         result = []
+        start = 0
+        values = None
+        if isinstance(self.element_strategy, SampledFromStrategy):
+            values = list(self.element_strategy.elements)
 
         while elements.more():
-            value = data.draw(self.element_strategy)
+            if values is None:
+                value = data.draw(self.element_strategy)
+            elif start < len(values):
+                # If we have a fixed number of values, use a variation on the
+                # shuffle from permutations() to consider each value at most
+                # once; avoiding the filter_too_much_healthcheck.
+                idx = integer_range(data, start, len(values) - 1)
+                value = values[idx]
+                values[start], values[idx] = values[idx], values[start]
+                start += 1
+            elif self.min_size <= len(result):
```

Comment:
Would this be better handled by setting a suitable `max_size` on the `more`? Breaking out of the more loop will confuse its invariants I think. In particular you won't get the terminating value that you're supposed to have in the byte stream to indicate the end of the sequence.

In general I'm a little suspicious about how well this will shrink. In particular it looks like it's going to really struggle to delete values from the generated list, because of the way the chosen value changes what is chosen later (I know this happens in `permutations` too, but I think the fact that we keep all the values in the list saves us in most cases because the order turns out to not matter very much)

For example, can we see how something like the following test does?

```python
@pytest.mark.parametrize('m, n', [
   (m, n) for n in range(2, 10)
   for m in range(n)
])
def test_shrink_for_inclusion(m, n):
    assert minimal(
        lists(sampled_from(range(n)), unique=True), lambda ls: m in ls
    ) == [m]
```

(we don't have to keep the full parametrization for the final result, I'm just not sure which values will be problematic and think it might be flaky).

My suspicion is that we're going to end up with a bunch of cases where that fails, although I'm not 100% certain of this and it would be pleasant to be wrong!

Reply:
Yep, setting the `max_size` (and turning the `average_size` into a property) handles that.

Unfortunately that test shows that it shrinks very, very badly.  This is fixable, but the simple version has quadratic worst-case complexity - comments below.

Reply:
Now feeling smug: by combining a set and a list of unused elements, it's gone from quadratic to amortised linear time 😄